### ISSUE TITLE
Deterministic Range Request Plugin Feature

### ISSUE THREAD

**Issue by ryandurfey** [apache/trafficserver#2662]
Create a new feature that allows caches to support fixed deterministic ranges, for example 1 MB blocks of content as individual objects.  If a client requests bytes 500,000-1,500,000 the cache would check cache for existing objects and if not in cache it would round this request to the deterministic boundaries and request any specific ranges that this request overlaps.  In this case it would request range 1-1,000,000 and 1,000,001 to 2,000,000 and use these to serve the client request.  

Logic:
1. Cache should check for existing cached objects or ranges to serve requests first
2. If missing any deterministic ranges that cover the client request the cache would fill those ranges as individually cached objects.
3. Cache fills client request based on content in these objects.

Additional Thoughts:
1. This should be configurable (off/on) on a per service (remap) basis
2. It would be nice to have the deterministic range size configurable on a per remap or at least global basis.
3. Should all files for a service be stored in this format for consistency or only files where ranges are requested first? For large files this could lead to many cache objects.
4. Should we fill the deterministic ranges on each call or using background fetch?

Background - Issues with the two current options for range request handling:
1. Background fetch of ranges
Issue: For very large files (>5 GB) this can be problematic,  blocking subsequent ranges during fill.  These can be unblocked, but it risks thundering herd issues at the origin.
Issue: It can lead to caching very large files when not needed like 100 GB software patches
2. Cache ranges as individual urls
Issue: this can be an issue if users send overlapping ranges which leads to caching the same content many times which could fill up cache.

**Comment by chassell** [comment]
This would be most useful if we had a desired 'sweet-spot' size of range request that was consistent.   If they arrived consistently and one of two (overlapping) blocks were ready to go for any reasonable size.

The ideal would be to get latency for any single-and-complete download way down without tons of duplicate requests.

**Comment by ashish-timilsina** [comment]
@ryandurfey 
@ashish-timilsina

**Comment by chassell** [comment]
@ashish-timilsina this project now has a working plugin.

It is mostly designed for small range requests but there is a large need for far bigger files being supported, possibly as a default.

**Comment by zwoop** [comment]
I believe this is fixed, reopen if not.

**Comment by mustafa1024m** [comment]
I asked about the plugin in the previous comment and got no response, now after search, I think it is the "Slice" plugin, and it requires "cache_range_requests" plugin, I found it here:
https://github.com/apache/trafficserver/tree/master/plugins/experimental/cache_range_requests

Please correct me if I am mistaken, but I think it does not exist in the documentation.

**Comment by ezelkow1** [comment]
> I asked about the plugin in the previous comment and got no response, now after search, I think it is the "Slice" plugin, and it requires "cache_range_requests" plugin, I found it here:
> https://github.com/apache/trafficserver/tree/master/plugins/experimental/cache_range_requests
> 
> Please correct me if I am mistaken, but I think it does not exist in the documentation.

Yes the slice plugin is the deterministic range slicing.  There are docs on master for it:
https://docs.trafficserver.apache.org/en/latest/admin-guide/plugins/slice.en.html

It is not in 8.x, however we are using it on 7.x right now (and have it in 8.x builds as well), so its possible to backport. Though I dont think we would be doing that officially since its a newer feature

**Comment by mustafa1024m** [comment]
Thanks of confirming, I actually read "This slice plugin, along with the cache_range_requests plugin", so I searched for the cache_range_requests, and I didn't find the docs for the cache_range_requests in the documentation (even in master), only in its GitHub page.
